### PR TITLE
Substitution of deprecated grep command/add_config_line for reading from/writing to glidein_config

### PR DESCRIPTION
**Pull request by namrathaurs**
Until now, `grep` was being used to read values from/write values to `glidein_config` when running system scripts during the glidein startup process. In 3.9.6, `add_config_line.source` was updated to include two utilities primarily to replace the grep-based logic in these scripts. The two functions, `gconfig_get` (for reading) and `gconfig_add` (for writing) will be used from release 3.9.7 onwards and therefore, the use of `grep`, for the purposes of reading and writing to `glidein_config` will be deprecated hereafter.

Note that, the `add_config_line.source` will need to be imported inside the script before either of the `gconfig_xxx` utilities can be invoked/used. It is only for this that `grep` and `cut` commands will continue to be used like before.

### CODE REVIEW

File: creation/web_base/condor_chirp [hunk 1]
```diff
@@ -9,7 +9,12 @@
 
 if [ -f "$1" ]; then
     glidein_config="$1"
-    error_gen=$(grep '^ERROR_GEN_PATH ' "$glidein_config" | awk '{print $2}')
+
+    # import add_config_line function
+    add_config_line_source=$(grep -m1 '^ADD_CONFIG_LINE_SOURCE ' "$glidein_config" | cut -d ' ' -f 2-)
+    source "$add_config_line_source"
+
+    error_gen=$(gconfig_get ERROR_GEN_PATH "$glidein_config")
```
Comment (mambelli) [inline creation/web_base/condor_chirp:17]: prefer `.` to `source`
If it is only for ERROR_GEN_PATH you can also keep the grep (1 occurrence and that variable is safe - not modified by the scripts)

File: creation/web_base/condor_platform_select.sh [hunk 2]
```diff
@@ -18,15 +18,15 @@
 glidein_config="$1"
 tmp_fname="${glidein_config}.$$.tmp"
 
-error_gen="`grep '^ERROR_GEN_PATH ' "$glidein_config" | cut -d ' ' -f 2-`"
+# import add_config_line function
+add_config_line_source=$(grep -m1 '^ADD_CONFIG_LINE_SOURCE ' "$glidein_config" | cut -d ' ' -f 2-)
+source "$add_config_line_source"
```
Comment (mambelli) [inline creation/web_base/condor_platform_select.sh:23]: Add the shellcheck line:
```
# shellcheck source=./add_config_line.source
```
prefer `.` to `source`

File: creation/web_base/create_temp_mapfile.sh [hunk 3]
```diff
@@ -15,10 +15,10 @@
 glidein_config="$1"
 
 # import add_config_line function
-add_config_line_source="`grep '^ADD_CONFIG_LINE_SOURCE ' "$glidein_config" | cut -d ' ' -f 2-`"
+add_config_line_source=$(grep -m1 '^ADD_CONFIG_LINE_SOURCE ' "$glidein_config" | cut -d ' ' -f 2-)
 source "$add_config_line_source"
```
Comment (mambelli) [inline creation/web_base/create_temp_mapfile.sh:19]: Add the shellcheck line:
```
# shellcheck source=./add_config_line.source
```
prefer `.` to `source`

File: creation/web_base/cvmfs_setup.sh [hunk 4]
```diff
@@ -6,16 +6,16 @@
...
 
-# fetch the error reporting helper script
-error_gen=$(grep '^ERROR_GEN_PATH ' $glidein_config | awk '{print $2}')
-
 # import add_config_line function
-add_config_line_source=$(grep '^ADD_CONFIG_LINE_SOURCE ' $glidein_config | awk '{print $2}')
+add_config_line_source=$(grep -m1 '^ADD_CONFIG_LINE_SOURCE ' "$glidein_config" | awk '{print $2}')
 # shellcheck source=./add_config_line.source
-. $add_config_line_source
+source $add_config_line_source
```
Comment (mambelli) [inline creation/web_base/cvmfs_setup.sh:12]: prefer . to source. Add quotes to the file name

**Comment by mambelli** [comment]
Generally when sourcing add_config_line:
- Make sure that the `$config_file` is quoted in the grep
- Add the shellcheck line:
- prefer `.` to `source`.
- Make sure that the file name is quoted.
```
add_config_line_source=$(grep -m1 '^ADD_CONFIG_LINE_SOURCE ' "$glidein_config" | awk '{print $2}')
# shellcheck source=./add_config_line.source
. "$add_config_line_source"
```

File: creation/web_base/get_id_selectors.source [hunk 5]
```diff
@@ -3,22 +3,24 @@
 # SPDX-FileCopyrightText: 2009 Fermi Research Alliance, LLC
 # SPDX-License-Identifier: Apache-2.0
 
+# NOTE: any script that imports this script to use the functions defined must also import add_config_line.source
```
Comment (mambelli) [inline creation/web_base/get_id_selectors.source:6]: Make sure that this is currently true. 
If you verified already ignore this comment

**Review by mambelli — CHANGES_REQUESTED** [review]
part 2

**Review by mambelli — DISMISSED** [review]

File: creation/web_base/cat_consts.sh [hunk 6]
```diff
@@ -20,20 +20,20 @@ function warn {
 }
 
 # import add_config_line function
-add_config_line_source="`grep '^ADD_CONFIG_LINE_SOURCE ' "$glidein_config" | cut -d ' ' -f 2-`"
+add_config_line_source=$(grep -m1 '^ADD_CONFIG_LINE_SOURCE ' "$glidein_config" | cut -d ' ' -f 2-)
 source "$add_config_line_source"
```
[the remark this answers is not in the document]
Comment (namrathaurs) [reply creation/web_base/cat_consts.sh:24]: done

File: creation/web_base/cat_consts.sh [hunk 7]
```diff
@@ -20,20 +20,20 @@ function warn {
...
 
 # import get_prefix function
-get_id_selectors_source="`grep '^GET_ID_SELECTORS_SOURCE ' "$glidein_config" | cut -d ' ' -f 2-`"
+get_id_selectors_source=$(grep -m1 '^GET_ID_SELECTORS_SOURCE ' "$glidein_config" | cut -d ' ' -f 2-)
```
[the remark this answers is not in the document]
Comment (namrathaurs) [reply creation/web_base/cat_consts.sh:27]: updated to use gconfig_get

File: creation/web_base/check_proxy.sh [hunk 8]
```diff
@@ -13,7 +13,7 @@
 glidein_config="$1"
 
 # import add_config_line function
-add_config_line_source="`grep '^ADD_CONFIG_LINE_SOURCE ' $glidein_config | cut -d ' ' -f 2-`"
+add_config_line_source=$(grep -m1 '^ADD_CONFIG_LINE_SOURCE ' $glidein_config | cut -d ' ' -f 2-)
 source "$add_config_line_source"
```
[the remark this answers is not in the document]
Comment (namrathaurs) [reply creation/web_base/check_proxy.sh:17]: updated

File: creation/web_base/check_blacklist.sh [hunk 9]
```diff
@@ -57,15 +57,19 @@ function check_blacklist {
...
-error_gen="`grep '^ERROR_GEN_PATH ' "$config_file" | cut -d ' ' -f 2-`"
+# import add_config_line function
+add_config_line_source=$(grep -m1 '^ADD_CONFIG_LINE_SOURCE ' "$config_file" | cut -d ' ' -f 2-)
+source "$add_config_line_source"
+
+error_gen=$(gconfig_get ERROR_GEN_PATH "$config_file")
 
 # import get_prefix function
-get_id_selectors_source="`grep '^GET_ID_SELECTORS_SOURCE ' "$config_file" | cut -d ' ' -f 2-`"
+get_id_selectors_source=$(grep -m1 '^GET_ID_SELECTORS_SOURCE ' "$config_file" | cut -d ' ' -f 2-)
```
[the remark this answers is not in the document]
Comment (namrathaurs) [reply creation/web_base/check_blacklist.sh:67]: updated

File: creation/web_base/check_blacklist.sh [hunk 10]
```diff
@@ -57,15 +57,19 @@ function check_blacklist {
 config_file="$1"
 dir_id=$2
 
-error_gen="`grep '^ERROR_GEN_PATH ' "$config_file" | cut -d ' ' -f 2-`"
+# import add_config_line function
+add_config_line_source=$(grep -m1 '^ADD_CONFIG_LINE_SOURCE ' "$config_file" | cut -d ' ' -f 2-)
+source "$add_config_line_source"
```
[the remark this answers is not in the document]
Comment (namrathaurs) [reply creation/web_base/check_blacklist.sh:62]: updated

File: creation/web_base/check_blacklist.sh [hunk 11]
```diff
@@ -57,15 +57,19 @@ function check_blacklist {
...
 
 # import get_prefix function
-get_id_selectors_source="`grep '^GET_ID_SELECTORS_SOURCE ' "$config_file" | cut -d ' ' -f 2-`"
+get_id_selectors_source=$(grep -m1 '^GET_ID_SELECTORS_SOURCE ' "$config_file" | cut -d ' ' -f 2-)
 source "$get_id_selectors_source"
 
 id_prefix=`get_prefix $dir_id`
 
-blacklist_file="`grep -i "^${id_prefix}BLACKLIST_FILE " "$config_file" | cut -d ' ' -f 2-`"
+blacklist_file=$(gconfig_get ${id_prefix}BLACKLIST_FILE "$config_file")
```
[the remark this answers is not in the document]
Comment (namrathaurs) [reply creation/web_base/check_blacklist.sh:72]: updated

File: creation/web_base/collector_setup.sh [hunk 12]
```diff
@@ -119,12 +119,12 @@ done | sort -n | cut -c8- | tr -d "\n" | sed -r 's/,+/,/g'`"
 
 _main() {
     # import add_config_line function
-    add_config_line_source="$(grep '^ADD_CONFIG_LINE_SOURCE ' "$glidein_config" | cut -d ' ' -f 2-)"
+    add_config_line_source=$(grep -m1 '^ADD_CONFIG_LINE_SOURCE ' "$glidein_config" | cut -d ' ' -f 2-)
     source "$add_config_line_source"
```
[the remark this answers is not in the document]
Comment (namrathaurs) [reply creation/web_base/collector_setup.sh:123]: updated

File: creation/web_base/condor_platform_select.sh (hunk 2 above)
Comment (namrathaurs) [reply creation/web_base/condor_platform_select.sh:23]: updated

File: creation/web_base/create_temp_mapfile.sh (hunk 3 above)
Comment (namrathaurs) [reply creation/web_base/create_temp_mapfile.sh:19]: updated

File: creation/web_base/singularity_wrapper.sh [hunk 13]
```diff
@@ -94,8 +94,12 @@ fi
 
 # TODO: Should I quit if glidein_config is not available?
 
+# import add_config_line function
+add_config_line_source=$(grep -m1 '^ADD_CONFIG_LINE_SOURCE ' $glidein_config | awk '{print $2}')
+source $add_config_line_source
+
 # error_gen defined also in singularity_lib.sh
-[[ -e "$glidein_config" ]] && error_gen="$(grep '^ERROR_GEN_PATH ' "$glidein_config" | cut -d ' ' -f 2-)"
+[[ -e "$glidein_config" ]] && error_gen=$(gconfig_get ERROR_GEN_PATH "$glidein_config")
```
[the remark this answers is not in the document]
Comment (namrathaurs) [reply creation/web_base/singularity_wrapper.sh:102]: reverted

File: creation/web_base/condor_chirp (hunk 1 above)
Comment (namrathaurs) [reply creation/web_base/condor_chirp:17]: reverted to using grep and added a comment line with the note about why grep is used still.

File: creation/web_base/glidein_startup.sh [hunk 14]
```diff
@@ -10,29 +10,6 @@
 # File Version:
 #
 
-# default IFS, to protect against unusual environment, better than "unset IFS" because works with restoring old one
-IFS=$' \t\n'
-
-global_args="$*"
```
[the remark this answers is not in the document]
Comment (namrathaurs) [reply creation/web_base/glidein_startup.sh]: The changes were intentional as I had modularized the existing code in this script independent of the changes that would go into v3.11. I was of the thought that making these drastic changes would not hurt/impact the later versions, but definitely did not think it through. I've made sure to revert the file to the latest commit's version of the file.

**Comment by namrathaurs** [comment]
All the requested changes have been completed and the comments have been marked as resolved.

File: creation/web_base/collector_setup.sh [hunk 15]
```diff
@@ -68,7 +68,7 @@ parse_and_select_collectors() {
     # Used for both User collectors and CCBs
     local inattr="$1"
 
-    local inlist="$(grep "^$inattr " "$glidein_config" | cut -d ' ' -f 2-)"
+    local inlist=$(gconfig_get $inattr "$glidein_config")
```
Comment (mambelli) [inline creation/web_base/collector_setup.sh:71]: quote `"$inattr"` to protect for spaces

File: creation/web_base/condor_platform_select.sh [hunk 16]
```diff
@@ -196,7 +197,7 @@ for version_el in $(echo "$condor_version" | tr ',' ' '); do
             condor_platform="${version_el}-${os_el}-${arch_el}"
             condor_platform_id="CONDOR_PLATFORM_$condor_platform"
 
-            condor_platform_check=`grep "^$condor_platform_id " "$glidein_config" | awk '{print $2}'`
+            condor_platform_check=$(gconfig_get $condor_platform_id "$glidein_config")
```
Comment (mambelli) [inline creation/web_base/condor_platform_select.sh:200]: quote variable `"$condor_platform_id"` to protect for spaces

File: creation/web_base/frontend/singularity_script_example.sh [hunk 17]
```diff
@@ -54,8 +54,13 @@ warn_raw() {
...
     glidein_config="$GWMS_THIS_SCRIPT_DIR/../glidein_config"
 
+# import add_config_line function
+add_config_line_source=$(grep -m1 '^ADD_CONFIG_LINE_SOURCE ' "$glidein_config" | awk '{print $2}')
+# shellcheck source=./add_config_line.source
+. "$add_config_line_source"
+
 # error_gen defined in singularity_lib.sh
-[[ -e "$glidein_config" ]] && error_gen=$(grep '^ERROR_GEN_PATH ' "$glidein_config" | cut -d ' ' -f 2-)
+[[ -e "$glidein_config" ]] && error_gen=$(gconfig_get ERROR_GEN_PATH "$glidein_config")
```
Comment (mambelli) [inline creation/web_base/frontend/singularity_script_example.sh:63]: This script may be invoked w/o passing glidein_config.
I would keep the grep for error_gen.
Alternatively, you have to protect the add_config_line lines (e.g. in an ` if [[ -e "$glidein_config" ]]...`)

**Review by mambelli — CHANGES_REQUESTED** [review]

File: creation/web_base/collector_setup.sh (hunk 15 above)
Comment (namrathaurs) [reply creation/web_base/collector_setup.sh:71]: updated

File: creation/web_base/condor_platform_select.sh (hunk 16 above)
Comment (namrathaurs) [reply creation/web_base/condor_platform_select.sh:200]: done

File: creation/web_base/frontend/singularity_script_example.sh (hunk 17 above)
Comment (namrathaurs) [reply creation/web_base/frontend/singularity_script_example.sh]: Thanks for sharing the suggestion; this has been reverted to use `grep` which was initially the case.

File: creation/web_base/frontend/singularity_script_example.sh [hunk 18]
```diff
@@ -54,6 +54,11 @@ warn_raw() {
 [[ -z "$glidein_config" ]] && [[ -e "$GWMS_THIS_SCRIPT_DIR/../glidein_config" ]] &&
     glidein_config="$GWMS_THIS_SCRIPT_DIR/../glidein_config"
 
+# import add_config_line function
+add_config_line_source=$(grep -m1 '^ADD_CONFIG_LINE_SOURCE ' "$glidein_config" | awk '{print $2}')
+# shellcheck source=./add_config_line.source
+. "$add_config_line_source"
+
```
Comment (mambelli) [inline creation/web_base/frontend/singularity_script_example.sh:61]: remove this part. 
"$glidein_config" could not exist
"$add_config_line_source" could be empty

**Review by mambelli — CHANGES_REQUESTED** [review]

File: creation/web_base/frontend/singularity_script_example.sh (hunk 18 above)
Comment (namrathaurs) [reply creation/web_base/frontend/singularity_script_example.sh]: done

**Review by mambelli — APPROVED** [review]

**Pull request merged**